src/NHibernate/Impl/DelayedAsyncEnumerator.cs

```diff
@@ -30,39 +30,40 @@ IAsyncEnumerator<T> IAsyncEnumerable<T>.GetEnumerator()
 
 		private sealed class Enumerator : IAsyncEnumerator<T>
 		{
-			private IEnumerable data;
-			private GetResult result;
-			private Delegate executeOnEval;
-			private IEnumerator enumerator;
+			private IEnumerable<T> _data;
+			private readonly GetResult _result;
+			private readonly Delegate _executeOnEval;
+			private IEnumerator<T> _enumerator;
 
 			public Enumerator(GetResult result, Delegate executeOnEval)
 			{
-				this.result = result;
-				this.executeOnEval = executeOnEval;
+				_result = result;
+				_executeOnEval = executeOnEval;
 			}
 
 			public T Current { get; private set; }
 
 			public async Task<bool> MoveNext(CancellationToken cancellationToken)
 			{
 				cancellationToken.ThrowIfCancellationRequested();
-				if (enumerator == null)
+				if (_enumerator == null)
 				{
-					data = await result();
-					if (executeOnEval != null)
-						data = (IEnumerable)executeOnEval.DynamicInvoke(data);
-					enumerator = data.GetEnumerator();
+					_data = await _result().ConfigureAwait(false);
+					if (_executeOnEval != null)
+						_data = (IEnumerable<T>)_executeOnEval.DynamicInvoke(_data);
```

Comment:
The execute on eval is fetched with post execute transformers, merged and compiled. Maybe I have missed a case but the one I have seen are either castable, or would anyway have failed at the now removed later `(T)` cast.